common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java

```diff
@@ -0,0 +1,238 @@
+        if (checkForFileExistence && !fileToWatch.toFile().exists()) {
+            throw new FileNotFoundException(String.format("File [%s] does not exist.", fileToWatch));
+        }
+
+        this.watchedFilePath = fileToWatch;
+        this.callback = callback;
+        setUncaughtExceptionHandler(uncaughtExceptionalHandler);
+
+        // Set the name for this object/thread for identification purposes.
+        this.setName("file-update-watcher-" + THREAD_NUM.incrementAndGet());
+        this.loopContinuously = loopContinuously;
+    }
+
+    @VisibleForTesting
+    String getWatchedFileName() {
+        Path fileName = this.watchedFilePath.getFileName();
+        if (fileName != null) {
+            return fileName.toString();
+        } else {
+            // It should never happen that file name is null, but we have to perform this defensive null check to
+            // satisfy SpotBugs.
+            throw new IllegalStateException("File name is null");
+        }
+    }
+
+    @VisibleForTesting
+    Path getWatchedDirectory() {
+        return this.watchedFilePath.getParent();
+    }
+
+    @Override
+    @SuppressWarnings("SleepWhileInLoop")
+    public void run() {
+        WatchKey watchKey = null;
+        WatchService watchService = null;
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+            log.debug("Done creating watch service for watching file at path: {}", this.watchedFilePath);
+
+            String fileName = getWatchedFileName();
+            Path directoryPath = getWatchedDirectory();
+            log.debug("Directory being watched is {}", directoryPath);
+
+            assert directoryPath != null;
+            directoryPath.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY,
+                    StandardWatchEventKinds.ENTRY_CREATE);
+            log.debug("Done setting up watch for modify entries for file at path: {}", this.watchedFilePath);
+
+            isWatchRegistered = true;
+
+            while (true) {
+                watchKey = watchService.take();
+                log.info("Retrieved and removed watch key for watching file at path: {}", this.watchedFilePath);
+
+                // Each file modification/create usually results in the WatcherService reporting the WatchEvent twice,
+                // as the file is updated twice: once for the content and once for the file modification time.
+                // These events occur in quick succession. We wait for 200 ms., here so that the events get
+                // de-duplicated - in other words only single event is processed.
+                //
+                // See https://stackoverflow.com/questions/16777869/java-7-watchservice-ignoring-multiple-occurrences-
+                // of-the-same-event for a discussion on this topic.
+                //
+                // If the two events are not raised within this duration, the callback will be invoked twice, which we
```

Comment:
The structure of this loop could be simplified a bit and prevent the issue which would avoid the confusion Fpj had below. If there were are private method factored out, this comment could be made into a javadoc.

Reply:
Sure, I've changed this part of the code. I first thought I should just take the comments and the `sleep` code out as you suggested, but then I was finding it hard to assign it a name that revealed its responsibility. So, I ended up extracting the watch key retrieval logic out too. The side effect, though, is that I couldn't take the comments out to method JavaDocs. Hope it looks OK.